From 77fb180458fde4a9ca7198984e62dfaa49ade4af Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
Date: Mon, 18 Feb 2013 22:29:48 +0100
Subject: [PATCH 124/715] drm/rcar-du: Don't re-reserve hardware plane at each
 update

Hardware planes only need to be reserved if not already reserved, or if
the number of required planes has changed. Release the previously
reserved planes (if any) and reserve new ones in that case.

Signed-off-by: Laurent Pinchart <laurent.pinchart+renesas@ideasonboard.com>
(cherry picked from commit c38cbeedb6536f8bf9bad0d17fcdceb9bf727583)

Signed-off-by: Simon Horman <horms+renesas@verge.net.au>
---
 drivers/gpu/drm/rcar-du/rcar_du_plane.c |   15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_plane.c b/drivers/gpu/drm/rcar-du/rcar_du_plane.c
index 110fd13..220ca00 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_plane.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_plane.c
@@ -206,6 +206,7 @@ rcar_du_plane_update(struct drm_plane *plane, struct drm_crtc *crtc,
 	struct rcar_du_plane *rplane = to_rcar_plane(plane);
 	struct rcar_du_device *rcdu = plane->dev->dev_private;
 	const struct rcar_du_format_info *format;
+	unsigned int nplanes;
 	int ret;
 
 	format = rcar_du_format_info(fb->pixel_format);
@@ -220,6 +221,8 @@ rcar_du_plane_update(struct drm_plane *plane, struct drm_crtc *crtc,
 		return -EINVAL;
 	}
 
+	nplanes = rplane->format ? rplane->format->planes : 0;
+
 	rplane->crtc = crtc;
 	rplane->format = format;
 	rplane->pitch = fb->pitches[0];
@@ -231,9 +234,15 @@ rcar_du_plane_update(struct drm_plane *plane, struct drm_crtc *crtc,
 	rplane->width = crtc_w;
 	rplane->height = crtc_h;
 
-	ret = rcar_du_plane_reserve(rplane);
-	if (ret < 0)
-		return ret;
+	/* Reallocate hardware planes if the number of required planes has
+	 * changed.
+	 */
+	if (format->planes != nplanes) {
+		rcar_du_plane_release(rplane);
+		ret = rcar_du_plane_reserve(rplane);
+		if (ret < 0)
+			return ret;
+	}
 
 	rcar_du_plane_compute_base(rplane, fb);
 	rcar_du_plane_setup(rplane);
-- 
1.7.10.4

